src/base/LruMap.h

```diff
@@ -135,63 +154,110 @@ LruMap<Key, EntryValue, EntryCost>::get(const Key &key)
     return NULL;
 }
 
-template <class Key, class EntryValue, size_t EntryCost>
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+LruMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t)
+{
+    // map the retention concept into a TTL value
+
+    // do not store / retain any entries
+    if (retention == 0)
+        return false;
+
+    // negative retention means store 'forever'
+    if (retention < 0)
+        return add(key, t, std::numeric_limits<int32_t>::max());
+
+    // positive retention are TTL
+    return add(key, t, retention);
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<Key, EntryValue, EntryCost>::add(const Key &key, EntryValue *t)
+LruMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t, int32_t ttl)
 {
-    if (ttl == 0)
+    if (retention == 0)
         return false;
 
     del(key);
-    trim();
 
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
-    storage.insert(MapPair(key, index.begin()));
+    // do not store expired() entries because findEntry() ignores
+    if (ttl <= 0)
+        return false;
+
+    auto e = new Entry(key, t, ttl);
+
+#if !SQUID_LRU_STRICT
+    // do not add to LRU at all if resulting memory use will
+    // knowingly exceed the memory limit and cause purging.
+    if (memUsed_ + e->memoryUsed() >= memLimit()) {
+        delete e;
+        return false;
+    } else
+#endif
+    {
+        memUsed_ += e->memoryUsed();
+        index.push_front(e);
 
-    ++entries_;
-    return true;
+        auto m = MapPair(key, index.begin());
+        memUsed_ += memoryUsedBy(m);
+        storage.insert(m);
+
+        ++entries_;
+
+        // if over memory limit, start purging
+        trim();
+    }
+
+    // the new entry was placed first,
+    // if trim removed it there will be none left.
+    return entries_ >= 1;
```

Comment:
Once we stop adding huge entries, we can simply return `true` here. We can assert that there are entries.